### PR TITLE
Add scaling tests

### PR DESCRIPTION
**Pull request by tarek-y-ismail**

### CODE REVIEW

File: mir-ci/mir_ci/tests/test_fractional_scale_v1.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        "fonts-ubuntu",
+        "adwaita-icon-theme-full",
+    ),
+    pip_pkgs=(
+        ("pygobject", "gi"),
+        ("robotframework~=6.1.1", "robot"),
+        ("rpaframework", "RPA"),
+        ("rpaframework-recognition", "RPA.recognition"),
+    ),
+)
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/test_fractional_scale_v1.py:68]: You want to modify this to fit your test - likely start with just `gtk4-examples` for deb.

**Review by Saviq — CHANGES_REQUESTED** [review]
Only looked at the CI results for now, but the direct cause for the failure is:
```
E               AssertionError: gtk4-demo died without being waited for or killed
```

But that's because it's just not installed (we should improve that error report to say the executable wasn't found).

**Comment by tarek-y-ismail** [comment]
> Only looked at the CI results for now, but the direct cause for the failure is:
> 
> ```
> E               AssertionError: gtk4-demo died without being waited for or killed
> ```
> 
> But that's because it's just not installed (we should improve that error report to say the executable wasn't found).

That's weird, I've been using it to test for the past two days. I recall seeing another error in the output of the robot near the top. Have you seen that?

**Comment by Saviq** [comment]
> That's weird, I've been using it to test for the past two days. I recall seeing another error in the output of the robot near the top. Have you seen that?

On your machine, yes - but who installed `gtk-4-examples` in CI? :)

File: mir-ci/mir_ci/tests/test_fractional_scale_v1.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    },
+)
+@pytest.mark.env(GSETTINGS_BACKEND="keyfile")
+@pytest.mark.parametrize("server", servers(ServerCap.FLOATING_WINDOWS | ServerCap.DISPLAY_CONFIG))
+@pytest.mark.deps(
+    debs=(
+        "libgirepository1.0-dev",
+        "libgtk-3-dev",
+        "fonts-ubuntu",
+        "adwaita-icon-theme-full",
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/test_fractional_scale_v1.py:60]:
```suggestion
        "gtk-4-examples",
```

File: mir-ci/mir_ci/tests/robot/suites/fractional_scale_v1/fractional_scale_v1.robot [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+Resource        ${KVM_RESOURCE}
+
+*** Variables ***
+${T}    ${CURDIR}
+
+*** Test Cases ***
+Click Button With Fractional Scaling Enabled
+    Log to console          ${KVM_RESOURCE}
+    Match                   ${T}/gtk4-demo-title-app-title.png
+    Move Pointer To         ${T}/gtk4-demo-button-simple-constraints.png
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/robot/suites/fractional_scale_v1/fractional_scale_v1.robot:11]: This keyword has embedded arguments:

https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#embedding-arguments-into-keyword-name

```suggestion
    Move Pointer To ${T}/gtk4-demo-button-simple-constraints.png
```

When you look at [the keyword documentation](https://github.com/canonical/mir-ci/blob/main/mir-ci/mir_ci/tests/robot/resources/kvm/kvm.resource), anywhere you see "$<foo>" in the keyword name itself, you don't separate it - that only works for those with `[Arguments]`.

**Comment by Saviq** [comment]
@tarek-y-ismail and with these tweaks your tests pass (the failures are only on neverputt, that's a known issue fixed in Mir already.

You can see the test results better in the "Test results" check. It sometimes gets attached to `PreCommit` for some reason...

On failure, you can also download the `test-results` archive from the "Test" workflow and dig for the appropriate "log.html" for your test - e.g. at the bottom here for a previous run:

https://github.com/canonical/mir-ci/actions/runs/10166013391

You'll then see a video of the run as well as the screenshot that failed to match - this way you can cut out exactly the template you need in CI.

All this... will get better soon when we integrate it all with the certification team's tooling.

**Review by Saviq — CHANGES_REQUESTED** [review]
I think you should be able to reduce the number of templates needed to just two per scale:
1. top-left portion of the app that fits at the scale in the minimum resolution
2. the expected window title after clicking the button

You can then use a static offset (multiplied by scale) to reach the intended button.

Updating the templates will require less work, then.

---

You got hung up on the "Combo Boxes" button, but at 1024x768 at scale >=1.75 it's not on screen when maximized, so maybe use/expect a different button? One that's on screen across all the cases you want to handle?

---

And I think you might have uncovered a bug in the initial placement logic - we probably should place the window's (0, 0) on screen, not it's lower right corner - when it doesn't fit.

File: mir-ci/mir_ci/tests/test_fractional_scale_v1.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+)
+class TestFractionalScaleV1:
+    async def test_fractional_scale_v1(self, robot_log, server, scale, tmp_path) -> None:
+        extensions = ("all",)  # TODO no need to enable all extension
+
+        server_instance = DisplayServer(
+            server,
+            add_extensions=extensions,
+            env={
+                "MIR_SERVER_X11_OUTPUT": "1024x768",
```
[the remark this answers is not in the document]
Comment (tarek-y-ismail) [reply mir-ci/mir_ci/tests/test_fractional_scale_v1.py:94]: X11 indeed defaults to 1280x1024 on my machine, but the CI defaults to 1024x768. I decided it would be easier to change the config on my machine (and any other person that might work on this suite later) to avoid future confusion. 

Alas, it turns out that the CI screenshots use logical scale instead of physical one, so this is not so useful now

**Comment by tarek-y-ismail** [comment]
> You got hung up on the "Combo Boxes" button, but at 1024x768 at scale >=1.75 it's not on screen when maximized, so maybe use/expect a different button? One that's on screen across all the cases you want to handle?

Regarding this, I wanted to aim for a button on a far edge of the application as that's where issues usually pop up when it comes to scaling, hence why I originally used "simple constraints". I worked around the button not being visible at higher scales by scrolling the list down and matching again if the button is not found 

> I think you should be able to reduce the number of templates needed to just two per scale:

I think I'll try this out. Thanks for the suggestion :)

**Comment by Saviq** [comment]
> Regarding this, I wanted to aim for a button on a far edge of the application as that's where issues usually pop up when it comes to scaling, hence why I originally used "simple constraints". I worked around the button not being visible at higher scales by scrolling the list down and matching again if the button is not found

I think it would be appropriate to just go for the last visible button, at the minimum supported resolution. You can also aim for different buttons on different scales.

File: mir-ci/mir_ci/tests/display_server_static_file.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import importlib
+import os
+import tempfile
+from typing import Any
+
+from mir_ci.program.display_server import DisplayServer
+
+
+class DisplayServerStaticFile:
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/display_server_static_file.py:9]: This can be undone now, not needed for this PR?

File: mir-ci/mir_ci/tests/robot/platforms/wayland/WaylandHid.py [hunk 6]
```diff
@@ -94,6 +98,15 @@ async def release_pointer_buttons(self) -> None:
...
             self.button(button, False)
 
+    @keyword
+    async def vertical_scroll(self, amount: float) -> None:
+        await self.connect()
+        if self.pointer:
+            self.pointer.axis_source(0)
+            self.pointer.axis(self.timestamp(), 0, amount)
+            self.pointer.frame()
+        self.display.roundtrip()
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/robot/platforms/wayland/WaylandHid.py:108]: I think this is excessive, we should be able to do without scrolling (just selecting buttons that are on screen at all times).

File: mir-ci/mir_ci/tests/robot/resources/kvm/kvm.resource [hunk 7]
```diff
@@ -212,3 +212,22 @@ Click ${button} Button
 Release Buttons
     [Documentation]    Release all buttons on the virtual pointer.
     HID.Release Pointer Buttons
+
+Scroll Until Match ${template}
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/robot/resources/kvm/kvm.resource:216]: Should be able to do without that, too.

File: mir-ci/mir_ci/tests/robot/suites/fractional_scale_v1/2.0-gtk4-demo-screenshot-floating.png
Comment (Saviq) [inline mir-ci/mir_ci/tests/robot/suites/fractional_scale_v1/2.0-gtk4-demo-screenshot-floating.png:1]: This, I think, shows a bug in Mir's initial placement logic. We should resize / maximize windows based on their initial/min size and available screen.

File: mir-ci/mir_ci/tests/test_display_configuration.py [hunk 8]
```diff
@@ -1,52 +1,15 @@
 import asyncio
-import importlib
-import os
-import tempfile
-from typing import Any
 from unittest.mock import ANY, Mock
 
 import pytest
+from display_server_static_file import DisplayServerStaticFile
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/test_display_configuration.py:5]: Ditto, undo this change here, please.

File: mir-ci/mir_ci/tests/test_fractional_scale_v1.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                        "card-id": 0,
+                        "unknown-1": {
+                            "state": "enabled",
+                            "scale": scale,
+                        },
+                    }
+                ]
+            }
+        }
+    }
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/test_fractional_scale_v1.py:55]: Unused, since you moved to `--display-scale`?

File: mir-ci/mir_ci/wayland/virtual_pointer.py [hunk 10]
```diff
@@ -53,23 +54,26 @@ def disconnected(self) -> None:
         pass
 
     def xdg_output_logical_size(self, xdg_output, width: int, height: int) -> None:
+        # Whenever the logical size changes, we updated the cached physical size
```
Comment (Saviq) [inline mir-ci/mir_ci/wayland/virtual_pointer.py:57]:
```suggestion
        # Whenever the logical size changes, we update the cached physical size
```

File: mir-ci/mir_ci/wayland/virtual_pointer.py [hunk 11]
```diff
@@ -53,23 +54,26 @@ def disconnected(self) -> None:
...
         if xdg_output == self.xdg_outputs[0]:
-            self.output_width = width
-            self.output_height = height
+            self.output_physical_width = int(width * self.output_scale)
+            self.output_physical_height = int(height * self.output_scale)
```
Comment (Saviq) [inline mir-ci/mir_ci/wayland/virtual_pointer.py:60]: I think I would rather stick to `output_width/height`, the only place where we have non-physical geometry is `xdg_output_logical_size`, so let's contain it there.

**Review by Saviq — CHANGES_REQUESTED** [review]
> I think it would be appropriate to just go for the last visible button, at the minimum supported resolution. You can also aim for different buttons on different scales.

Ping on this one.

File: mir-ci/mir_ci/tests/test_fractional_scale_v1.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Saviq) [reply mir-ci/mir_ci/tests/test_fractional_scale_v1.py:76]: This needs to go, then.

File: mir-ci/mir_ci/tests/robot/platforms/wayland/WaylandHid.py [hunk 12]
```diff
@@ -26,6 +26,10 @@ def __init__(self) -> None:
         display_name = os.environ.get("WAYLAND_DISPLAY", "wayland-0")
         super().__init__(display_name)
 
+    @keyword
+    async def set_output_scale(self, output_scale: float) -> None:
+        self.output_scale = output_scale
+
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/robot/platforms/wayland/WaylandHid.py:32]: This should go.

File: mir-ci/mir_ci/tests/display_server_static_file.py (hunk 5 above)
Comment (Saviq) [reply mir-ci/mir_ci/tests/display_server_static_file.py:9]: This file still needs to go.

File: mir-ci/mir_ci/wayland/virtual_pointer.py [hunk 13]
```diff
@@ -53,9 +54,10 @@ def disconnected(self) -> None:
         pass
 
     def xdg_output_logical_size(self, xdg_output, width: int, height: int) -> None:
+        # Whenever the logical size changes, we update the cached physical size
```
Comment (Saviq) [inline mir-ci/mir_ci/wayland/virtual_pointer.py:57]: We should probably just replace this by [wl_output::geometry](https://wayland.app/protocols/wayland#wl_output:event:geometry), we can then avoid thinking of scale here altogether.

But that's for a separate PR.

File: mir-ci/mir_ci/tests/robot/suites/fractional_scale_v1/fractional_scale_v1.robot [hunk 14]
```diff
@@ -0,0 +1,46 @@
+*** Settings ***
+Resource        ${KVM_RESOURCE}
+
+Test Setup      Set Output Scale    ${SCALE}
+
+
+*** Variables ***
+${T}                        ${CURDIR}
+${FLOATING-SCREENSHOT}      ${T}/${SCALE}-gtk4-demo-screenshot-floating.png
+${MAXIMIZED-SCREENSHOT}     ${T}/${SCALE}-gtk4-demo-screenshot-maximized.png
```
Comment (Saviq) [inline mir-ci/mir_ci/tests/robot/suites/fractional_scale_v1/fractional_scale_v1.robot:10]: Unused, can be removed, along with the templates.

**Review by Saviq — CHANGES_REQUESTED** [review]
Handful of nits.

I'm looking into why they're failing in CI - there's some subtle differences, we may want to avoid having the full screenshots as templates, as e.g. Frame will always maximize.

File: mir-ci/mir_ci/tests/display_server_static_file.py (hunk 5 above)
Comment (tarek-y-ismail) [reply mir-ci/mir_ci/tests/display_server_static_file.py:9]: I think I removed it in a previous commit but your force push overwrote that. Should be easy enough to redo :)

Comment (tarek-y-ismail) [reply mir-ci/mir_ci/tests/display_server_static_file.py:9]: Nevermind, I actually forgot to remove the file

File: mir-ci/mir_ci/tests/robot/suites/fractional_scale_v1/2.0-gtk4-demo-screenshot-floating.png
Comment (Saviq) [reply mir-ci/mir_ci/tests/robot/suites/fractional_scale_v1/2.0-gtk4-demo-screenshot-floating.png:1]: canonical/mir#3553

**Review by Saviq — APPROVED** [review]

**Comment by tarek-y-ismail** [comment]
Additional changes  reviewed, LGTM! Will just rebase to clean up the commit history then merge.
